### PR TITLE
Review and suggested improvements

### PR DESCRIPTION
**Pull request by sehe**
- Doin' a portability
- Review and suggested improvements

### CODE REVIEW

File: library.cpp [hunk 1]
```diff
@@ -38,33 +38,66 @@ boost::mutex mtx_;
...
     WideCharToMultiByte(CP_UTF8, 0, &wstr[0], (int)wstr.size(), &strTo[0], size_needed, NULL, NULL);
     return strTo;
+#else
+    std::mbstate_t state{};
+
+    auto in         = wstr.c_str();
+    int size_needed = std::wcsrtombs(nullptr, &in, wstr.size(), &state);
+
+    std::string strTo(1 + size_needed, 0);
+    size_t n = std::wcsrtombs(strTo.data(), &in, wstr.size(), &state);
```
Comment (sehe) [inline library.cpp]: note this was fixed in the [subsequent commit](https://github.com/Danp2/WebSocketAsio/pull/1/commits/e8c6b02347b8c1dcf36fe97a7f1ea708495b07a2#diff-7f6369d1fe1e7e2a43cb0edaeeb0de3517327139a9fce593f0fda4f22c3929cbR72). I should probably have cleaned up for a PR, then again, the PR is mainly the most convenient way to share the review, not so much intended as verbatim contribution

**Comment by Danp2** [comment]
@sehe Thanks for taking a look and suggesting improvements. The changes to library.cpp are substantial so it may take me a while to review those and also evaluate how they will affect my pending changes in https://github.com/Danp2/WebSocketAsio/tree/dp_update. I'm sure that I will have some questions for you in the near future. 😉

**Comment by Danp2** [comment]
Hi @sehe 

> There was very big OOPS with the buffer passed to async_write being
destructed at the end of `send_message` - so that was undefined
behaviour. 

Can you elaborate on this and explain which change in your code addresses it?

> Regardless, you didn't make sure that no overlapping writes
were possible. I inserted the outgoing queue pattern (`_outbox`).

I had already implemented this in my other branch.

File: library.cpp [hunk 2]
```diff
@@ -1,518 +1,403 @@
...
-    session(net::io_context &ioc)
-            : resolver_(net::make_strand(ioc)), ws_(net::make_strand(ioc)), is_first_write_(false) {
-    }
-    /// Send message to remote websocket server
-    /// \param data to be sent
-    void send_message(const std::wstring &data) {
-        if(EnableVerbose)
-            std::wcout << L"<WsDll-" ARCH_LABEL ">> Sending message: " << data << std::endl;
-
-      const std::string to_send = utf8_encode(data);
```
Comment (sehe) [inline library.cpp]: this is a local variable, it doesn't live long enough for `async_write` to complete

**Comment by sehe** [comment]
> Can you elaborate on this and explain which change in your code addresses it?

Here https://github.com/Danp2/WebSocketAsio/pull/1/commits/e8c6b02347b8c1dcf36fe97a7f1ea708495b07a2#diff-7f6369d1fe1e7e2a43cb0edaeeb0de3517327139a9fce593f0fda4f22c3929cbL126 you had a local variable that is destructed at the end of the function scope, so before the completion of `async_write`.

My initial fix was to make it a member variable, but you can't see it because it was superseded by the - also required - changes:

 - to make and outbound queue (`outbox_`)
 - splitting `send_message`, `disconnect` etc from `do_send_message` and `do_disconnect` to make sure they run on the strand (or, really, the service thread in the first place).

I just noticed that I omitted the vital change to `send_message` analogous to `disconnect`:

```c++
void send_message(std::wstring data)
{
    post(ws_.get_executor(),
         std::bind(&Session::do_send_message, shared_from_this(), std::move(data)));
}
```

File: library.cpp [hunk 3]
```diff
@@ -137,13 +137,10 @@ namespace /*anon*/ {
...
         /// \param data to be sent
-        void send_message(std::wstring const& data)
+        void send_message(std::wstring data)
         {
-            VERBOSE(L"Sending message: " << data);
-
-            const std::string to_send = utf8_encode(data); // BIG OOPS! lifetime ends at function exit
-            ws_.async_write(net::buffer(to_send),
-                            beast::bind_front_handler(&Session::on_write, shared_from_this()));
+            post(ws_.get_executor(),
```
Comment (sehe) [inline library.cpp:142]: added the missed spot - hopefully it was already clear from the analogous change to `disconnect`/`do_disconnect` :)

**Comment by sehe** [comment]
> I had already implemented this in my other branch.

w.r.t to the "surprise" branch I'm going to leave that to you: after all, my journey to the GH repository was already a "chase" from the SO question, so I think you'll be able to manage the rest (also, it just shows how much it helps to give enough context in the question).

File: library.cpp (hunk 3 above)
Comment (Danp2) [reply library.cpp:142]: Actually, it wasn't. Hence my questions. It makes more sense now, thanks!

**Comment by Danp2** [comment]
> w.r.t to the "surprise" branch

I actually linked to this branch in my post on SO, but no worries. Your help is greatly appreciated.

**Comment by sehe** [comment]
> > w.r.t to the "surprise" branch
> 
> I actually linked to this branch in my post on SO, but no worries. Your help is greatly appreciated.

You're quite right then. Apparently I'm just very bad at UIs :) I couldn't figure out it was showing a branch (`tree` reminded me only of git `tree`s which are literally part and parcel of every commit, e.g. when viewing a subtree). I just truncated the URL to something that `git clone` would work with and thought I had it :)

Cheers! By the way, if you want I can show my idea around multi-session implementation. It actually becomes simpler in many ways and a lot more unsurprising to the end-user. The end-user can still decide to put the session handle in a AHK global variable if they don't care about multi-session capabilities.

The reason I didn't jump to it is because I'm not sure how to marshal void* over to AHK (I suspect `IntPtr` will do, but I have no installations to actually test with)

**Comment by Danp2** [comment]
Sure, I'd like that. Would it make sense to merge this PR first and then create a new PR? That way I could focus on reviewing my other pending changes.

**Comment by Danp2** [comment]
I don't use AHK either. I'm intending to use it with AutoIt, which is an AHK competitor. Hopefully we can make it work for both and everyone benefits.

**Pull request merged**